protokt-reflect/src/jvmMain/kotlin/protokt/v1/google/protobuf/KtMessages.kt

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (c) 2023 Toast, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:JvmName("KtMessages")
+
+package protokt.v1.google.protobuf
+
+import com.google.protobuf.DescriptorProtos
+import com.google.protobuf.Descriptors
+import com.google.protobuf.Descriptors.FieldDescriptor
+import com.google.protobuf.Descriptors.FieldDescriptor.Type
+import com.google.protobuf.DynamicMessage
+import com.google.protobuf.MapEntry
+import com.google.protobuf.Message
+import com.google.protobuf.UnknownFieldSet
+import com.google.protobuf.UnknownFieldSet.Field
+import com.google.protobuf.UnsafeByteOperations
+import com.google.protobuf.WireFormat
+import com.toasttab.protokt.v1.ProtoktProtos
+import io.github.classgraph.ClassGraph
+import protokt.v1.Bytes
+import protokt.v1.Converter
+import protokt.v1.KtEnum
+import protokt.v1.KtGeneratedFileDescriptor
+import protokt.v1.KtGeneratedMessage
+import protokt.v1.KtMessage
+import protokt.v1.reflect.ClassLookup
+import protokt.v1.reflect.FieldType
+import protokt.v1.reflect.WellKnownTypes
+import protokt.v1.reflect.inferClassName
+import protokt.v1.reflect.resolvePackage
+import protokt.v1.reflect.typeName
+import kotlin.Any
+import kotlin.reflect.KClass
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.reflect.full.findAnnotation
+
+fun KtMessage.toDynamicMessage(context: RuntimeContext): DynamicMessage =
+    context.convertValue(this) as DynamicMessage
+
+fun KtMessage.hasField(field: FieldDescriptor): Boolean {
+    val value = getField(field)
+
+    return if (field.hasPresence()) {
+        value != null
+    } else {
+        value != defaultValue(field)
+    }
+}
+
+private fun defaultValue(field: FieldDescriptor) =
+    when (field.type) {
+        Type.UINT64, Type.FIXED64 -> 0uL
+        Type.UINT32, Type.FIXED32 -> 0u
+        Type.BYTES -> Bytes.empty()
+        else -> field.defaultValue
+    }
+
+fun KtMessage.getField(field: FieldDescriptor) =
+    ProtoktReflect.getField(this, field)
+
+class RuntimeContext internal constructor(
+    descriptors: Iterable<Descriptors.Descriptor>,
+    private val classLookup: ClassLookup
+) {
+    constructor(descriptors: Iterable<Descriptors.Descriptor>) : this(descriptors, DEFAULT_CLASS_LOOKUP)
+
+    internal val descriptorsByTypeName = descriptors.associateBy { it.fullName }
+
+    fun convertValue(value: Any?) =
```

Comment:
not obvious what this public API method does

Reply:
I can make this not take a nullable argument for starters.

Would naming it better be a good idea? I think it has to be able to take any type since it's called at runtime with whatever field value the descriptor extracts from a protokt message.

A more descriptive name could be `convertProtoktToProtobufJava`.

Reply:
The reason this has to be public is to enable [sub-message conversion in protovalidate](https://github.com/andrewparmet/protovalidate-protokt/blob/protokt-v2/src/main/java/build/buf/protovalidate/internal/evaluator/ProtoktImpl.kt). If it's acceptable to convert the whole message every time, even if some portion of the message may never be validated, then this can be internal.